driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -635,14 +648,97 @@ public void shutdown() {
             // we're not on an I/O thread or anything, so just call await.
             allChannels.close().awaitUninterruptibly();
 
-            // This will call shutdownNow on the boss and worker executor. Since this is called
-            // only once all connection have been individually closed, it's fine.
-            channelFactory.releaseExternalResources();
+            // This will shut down I/O threads. Since this is called only once all connection have
+            // been individually closed, it's fine.
+            eventLoopGroup.shutdownGracefully().syncUninterruptibly();
             timer.stop();
         }
     }
 
-    private class Dispatcher extends IdleStateAwareChannelUpstreamHandler {
+    private static final class Flusher implements Runnable {
+        final EventLoop eventLoop;
+        final ConcurrentLinkedQueue<FlushItem> queued = new ConcurrentLinkedQueue<FlushItem>();
+        final AtomicBoolean running = new AtomicBoolean(false);
+        final HashSet<Channel> channels = new HashSet<Channel>();
```

Comment:
Hard to come up with a one-size-fits-all for this one. There is one Flusher per EventLoop, so the max size of this set equals the number of Channels per EventLoop. This depends on the EventLoopGroup configuration (which we'll make pluggable with JAVA-676) and the total number of connections (which depend on number of hosts, pool configuration...).